promql/engine.go

```diff
@@ -668,16 +691,22 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 
 // rangeEval evaluates the given expressions, and then for each step calls
 // the given function with the values computed for each expression at that
-// step.  The return value is the combination into time series of  of all the
+// step.  The return value is the combination into time series of all the
 // function call results.
 func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ...Expr) Matrix {
 	numSteps := int((ev.endTimestamp-ev.startTimestamp)/ev.interval) + 1
 	matrixes := make([]Matrix, len(exprs))
 	origMatrixes := make([]Matrix, len(exprs))
+	tempNumSamples := 0
```

Comment:
Shouldn't this be ev.currentSamples? Ultimately when this function returns ev.currentSamples should be the previous value, plus how ever many samples we're returning.

Reply:
wouldn't that just be whatever the value of `ev.currentSamples` is when we return?

Reply:
In every case? After the call to eval, we've potentially added to `ev.currentSamples` in  eval via`*Call`, `*VectorSelector` cases, or via `vectorSelector()`/`matrixSelector()` in which cases we wouldn't have reset `ev.currentSamples` afterwards, or we've called `rangeEval()`.

As I said earlier, I'm sill a bit lost understanding how to properly handle all of the cases. Any pointers would be helpful.

Reply:
Yes, in every case. 

> After the call to eval, we've potentially added to ev.currentSamples in ...

Yes, and that needs to be undone before we return as those are no longer in memory.

So if we start a rangeEval with say 10 samples used up by higher in the call stack, then any functions we call we will add to that - so say we're now at 30 samples. If the output of the rangeEval is 5 samples then currentSamples should be 15 by time the rangeEval returns.

Reply:
so then here before a call to `ev.eval` we want to be sure `ev.currentSamples` is saved to a temp var, and after the call `ev.currentSamples` should be set to that temp var plus the length of all the series in `matrixes[i]`?

Reply:
Yes